source/addonHandler/__init__.py

```diff
@@ -40,42 +37,77 @@
 ADDON_PENDINGINSTALL_SUFFIX=".pendingInstall"
 DELETEDIR_SUFFIX=".delete"
 
-state={}
-
 # Add-ons that are blocked from running because they are incompatible
 _blockedAddons=set()
 
-def loadState():
-	global state
-	statePath=os.path.join(globalVars.appArgs.configPath,stateFilename)
-	try:
-		# #9038: Python 3 requires binary format when working with pickles.
-		with open(statePath, "rb") as f:
-			state = pickle.load(f)
-		if "disabledAddons" not in state:
-			state["disabledAddons"] = set()
-		if "pendingDisableSet" not in state:
-			state["pendingDisableSet"] = set()
-		if "pendingEnableSet" not in state:
-			state["pendingEnableSet"] = set()
-	except:
-		# Defaults.
-		state = {
-			"pendingRemovesSet":set(),
-			"pendingInstallsSet":set(),
-			"disabledAddons":set(),
-			"pendingEnableSet":set(),
-			"pendingDisableSet":set(),
-		}
+
+class AddonsState(collections.UserDict):
+
+	_DEFAULT_STATE_CONTENT = {
+		"pendingRemovesSet": set(),
+		"pendingInstallsSet": set(),
+		"disabledAddons": set(),
+		"pendingEnableSet": set(),
+		"pendingDisableSet": set(),
+	}

```

Comment:
`SpeechState` in [speech.py](https://github.com/nvaccess/nvda/blob/master/source/speech/speech.py#L77), uses a dataclass. That way we have typed information on the items in the state that the rest of NVDA can use.

If there isn't a reason to specifically subclass `collections.UserDict`, it could be possible to just extend the dataclass with these functions, or replace the `UserDict` usage with a `self.state` variable. To convert from pickled dictionaries, dataclasses can be [converted to](https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict), and [updated from](https://docs.python.org/3/library/dataclasses.html#dataclasses.replace) dictionaries.

Reply:
A reason to keep using `collections.UserDict` is to maintain backward compatibility with `state`. I think this should be documented under the class declaration if that is the case.

Reply:
Is preserving backward compatibility enough of a reason to keep subclassing `collections.UserDict` in your opinion?

Reply:
I've added an explanatory comment.

Reply:
Yes, it's just helpful to document why the usage is different to other parts of the codebase. I also wasn't sure if the intention was to change the API.